File: reference-apps/nodejs/src/routes/redis-cluster.js
/**
 * Redis Cluster Information API endpoints
 *
 * Provides detailed information about Redis cluster configuration:
 * - Cluster nodes and their assigned slots
 * - Slot distribution
 * - Cluster topology
 * - Per-node detailed information
 */

const express = require('express');
const redis = require('redis');
const { logger } = require('../middleware/logging');
const { getVaultSecret } = require('../services/vault');
const config = require('../config');

const router = express.Router();

/**
 * Get cluster nodes information
 */
router.get('/cluster/nodes', async (req, res) => {
  try {
    // Get Redis credentials from Vault
    const creds = await getVaultSecret('redis-1');
    const password = creds.password;

    const client = redis.createClient({
      socket: {
        host: config.redis.host,
        port: config.redis.port,
        connectTimeout: 5000
      },
      password,
    });

    await client.connect();

    // Get cluster nodes information
    const clusterNodesRaw = await client.sendCommand(['CLUSTER', 'NODES']);
    await client.quit();

    // Parse the output
    const nodes = [];
    const lines = clusterNodesRaw.split('\n');

    for (const line of lines) {
      if (!line.trim()) continue;

      const parts = line.split(' ');
      if (parts.length < 8) continue;

      const nodeId = parts[0];
      const address = parts[1];
      const flags = parts[2];
      const masterId = parts[3] !== '-' ? parts[3] : null;
      const pingSent = parts[4];
      const pongRecv = parts[5];
      const configEpoch = parts[6];
      const linkState = parts[7];

      // Parse slots
      const slotRanges = [];
      for (let i = 8; i < parts.length; i++) {
        const slotInfo = parts[i];
        if (slotInfo.includes('-') && !slotInfo.startsWith('[')) {
          const [start, end] = slotInfo.split('-');
          slotRanges.push({ start: parseInt(start), end: parseInt(end) });
        } else if (/^\d+$/.test(slotInfo)) {
          slotRanges.push({ start: parseInt(slotInfo), end: parseInt(slotInfo) });
        }
      }

      // Parse address
      const hostPort = address.split('@')[0];
      const lastColon = hostPort.lastIndexOf(':');
      const host = hostPort.substring(0, lastColon);
      const port = parseInt(hostPort.substring(lastColon + 1));

      // Determine role
      let role = 'unknown';
      if (flags.includes('master')) role = 'master';
      else if (flags.includes('slave')) role = 'replica';

      nodes.push({
        node_id: nodeId,
        host,
        port,
        role,
        flags: flags.split(','),
        master_id: masterId,
        ping_sent: pingSent,
        pong_recv: pongRecv,
        config_epoch: configEpoch,
        link_state: linkState,
        slot_ranges: slotRanges,
        slots_count: slotRanges.reduce((sum, range) => sum + (range.end - range.start + 1), 0)
      });
    }

    res.json({
      total_nodes: nodes.length,
      nodes,
      raw: clusterNodesRaw
    });
  } catch (error) {
    logger.error('Error fetching cluster nodes', { error: error.message, requestId: req.requestId });
    res.status(500).json({
      error: 'Failed to fetch cluster nodes',
      message: error.message
    });
  }
});

/**
 * Get cluster slots information
 */
router.get('/cluster/slots', async (req, res) => {
  try {
    const creds = await getVaultSecret('redis-1');
    const password = creds.password;

    const client = redis.createClient({
      socket: {
        host: config.redis.host,
        port: config.redis.port,
        connectTimeout: 5000
      },
      password,
    });

    await client.connect();

    const clusterSlotsRaw = await client.sendCommand(['CLUSTER', 'SLOTS']);
    await client.quit();

    res.json({
      slots: clusterSlotsRaw,
      total_slot_ranges: Array.isArray(clusterSlotsRaw) ? clusterSlotsRaw.length : 0
    });
  } catch (error) {
    logger.error('Error fetching cluster slots', { error: error.message, requestId: req.requestId });
    res.status(500).json({
      error: 'Failed to fetch cluster slots',
      message: error.message
    });
  }
});

/**
 * Get cluster info
 */
router.get('/cluster/info', async (req, res) => {
  try {
    const creds = await getVaultSecret('redis-1');
    const password = creds.password;

    const client = redis.createClient({
      socket: {
        host: config.redis.host,
        port: config.redis.port,
        connectTimeout: 5000
      },
      password,
    });

    await client.connect();

    const clusterInfoRaw = await client.sendCommand(['CLUSTER', 'INFO']);
    await client.quit();

    // Parse cluster info
    const info = {};
    const lines = clusterInfoRaw.split('\r\n');
    for (const line of lines) {
      if (line.includes(':')) {
        const [key, value] = line.split(':');
        info[key] = value;
      }
    }

    res.json({
      cluster_info: info,
      raw: clusterInfoRaw
    });
  } catch (error) {
    logger.error('Error fetching cluster info', { error: error.message, requestId: req.requestId });
    res.status(500).json({
      error: 'Failed to fetch cluster info',
      message: error.message
    });
  }
});

/**
 * Get detailed information for a specific node
 */
router.get('/nodes/:node_name/info', async (req, res) => {
  try {
    const { node_name } = req.params;
    const creds = await getVaultSecret('redis-1');
    const password = creds.password;

    const client = redis.createClient({
      socket: {
        host: node_name,
        port: 6379,
        connectTimeout: 5000
      },
      password,
    });

    await client.connect();

    const serverInfo = await client.info();
    await client.quit();

    // Parse INFO response
    const sections = {};
    let currentSection = 'general';
    sections[currentSection] = {};

    const lines = serverInfo.split('\r\n');
    for (const line of lines) {
      if (line.startsWith('#')) {
        currentSection = line.substring(2).toLowerCase().trim();
        sections[currentSection] = {};
      } else if (line.includes(':')) {
        const [key, value] = line.split(':');
        sections[currentSection][key] = value;
      }
    }

    res.json({
      node: node_name,
      sections,
      raw: serverInfo
    });
  } catch (error) {
    logger.error('Error fetching node info', { node: req.params.node_name, error: error.message, requestId: req.requestId });
    res.status(500).json({
      error: 'Failed to fetch node info',
      message: error.message,
      node: req.params.node_name
    });
  }
});

module.exports = router;

File: reference-apps/fastapi/app/routers/redis_cluster.py
"""
Redis Cluster Information API endpoints

Provides detailed information about Redis cluster configuration:
- Cluster nodes and their assigned slots
- Slot distribution
- Cluster topology
- Per-node detailed information
"""

from fastapi import APIRouter
import redis.asyncio as redis
import logging

from app.config import settings
from app.services.vault import vault_client

logger = logging.getLogger(__name__)
router = APIRouter()


@router.get("/cluster/nodes")
async def get_cluster_nodes():
    """
    Get detailed information about all Redis cluster nodes

    Returns information including:
    - Node ID
    - IP address and port
    - Flags (master/slave)
    - Master node ID (for replicas)
    - Ping/pong timestamps
    - Config epoch
    - Link state
    - Slots assigned to each node
    """
    try:
        # Fetch credentials from Vault
        creds = await vault_client.get_secret("redis-1")
        password = creds.get("password")

        # Connect to first node to get cluster info
        client = redis.Redis(
            host=settings.REDIS_HOST,
            port=settings.REDIS_PORT,
            password=password,
            decode_responses=True,
            socket_connect_timeout=5
        )

        # Get cluster nodes information
        cluster_nodes_raw = await client.execute_command("CLUSTER", "NODES")
        await client.close()

        # Parse CLUSTER NODES output
        nodes = []
        for line in cluster_nodes_raw.strip().split("\n"):
            if not line:
                continue

            parts = line.split()
            if len(parts) < 8:
                continue

            node_id = parts[0]
            address = parts[1]
            flags = parts[2]
            master_id = parts[3] if parts[3] != "-" else None
            ping_sent = parts[4]
            pong_recv = parts[5]
            config_epoch = parts[6]
            link_state = parts[7]

            # Parse slots (if any)
            slots = []
            slot_ranges = []
            for i in range(8, len(parts)):
                slot_info = parts[i]
                if "-" in slot_info and not slot_info.startswith("["):
                    # Slot range like "0-5460"
                    start, end = slot_info.split("-")
                    slot_ranges.append({"start": int(start), "end": int(end)})
                    slots.extend(range(int(start), int(end) + 1))
                elif slot_info.isdigit():
                    # Single slot
                    slots.append(int(slot_info))
                    slot_ranges.append({"start": int(slot_info), "end": int(slot_info)})

            # Parse address
            host_port = address.split("@")[0]  # Remove cluster bus port
            host, port = host_port.rsplit(":", 1)

            # Determine role
            role = "master" if "master" in flags else "replica" if "slave" in flags else "unknown"

            nodes.append({
                "node_id": node_id,
                "host": host,
                "port": int(port),
                "role": role,
                "flags": flags.split(","),
                "master_id": master_id,
                "ping_sent": ping_sent,
                "pong_recv": pong_recv,
                "config_epoch": int(config_epoch),
                "link_state": link_state,
                "slots_count": len(slots),
                "slot_ranges": slot_ranges
            })

        return {
            "status": "success",
            "total_nodes": len(nodes),
            "nodes": nodes
        }

    except Exception as e:
        logger.error(f"Failed to get cluster nodes: {e}")
        return {"status": "error", "error": "Failed to retrieve cluster nodes information"}


@router.get("/cluster/slots")
async def get_cluster_slots():
    """
    Get slot distribution across cluster nodes

    Returns:
    - Slot ranges assigned to each master
    - Total slots covered
    - Slot coverage percentage
    """
    try:
        # Fetch credentials from Vault
        creds = await vault_client.get_secret("redis-1")
        password = creds.get("password")

        # Connect to first node
        client = redis.Redis(
            host=settings.REDIS_HOST,
            port=settings.REDIS_PORT,
            password=password,
            decode_responses=False,  # CLUSTER SLOTS returns binary data
            socket_connect_timeout=5
        )

        # Get cluster slots information
        cluster_slots = await client.execute_command("CLUSTER", "SLOTS")
        await client.close()

        # Parse CLUSTER SLOTS output
        slot_distribution = []
        total_slots = 0

        for slot_info in cluster_slots:
            start_slot = slot_info[0]
            end_slot = slot_info[1]
            master_info = slot_info[2]

            # Master node info
            master_host = master_info[0].decode() if isinstance(master_info[0], bytes) else master_info[0]
            master_port = master_info[1]
            master_id = master_info[2].decode() if isinstance(master_info[2], bytes) else master_info[2]

            # Replica info (if any)
            replicas = []
            for i in range(3, len(slot_info)):
                replica_info = slot_info[i]
                replica_host = replica_info[0].decode() if isinstance(replica_info[0], bytes) else replica_info[0]
                replica_port = replica_info[1]
                replica_id = replica_info[2].decode() if isinstance(replica_info[2], bytes) else replica_info[2]
                replicas.append({
                    "host": replica_host,
                    "port": replica_port,
                    "node_id": replica_id
                })

            slots_in_range = end_slot - start_slot + 1
            total_slots += slots_in_range

            slot_distribution.append({
                "start_slot": start_slot,
                "end_slot": end_slot,
                "slots_count": slots_in_range,
                "master": {
                    "host": master_host,
                    "port": master_port,
                    "node_id": master_id
                },
                "replicas": replicas
            })

        return {
            "status": "success",
            "total_slots": total_slots,
            "max_slots": 16384,
            "coverage_percentage": round((total_slots / 16384) * 100, 2),
            "slot_distribution": slot_distribution
        }

    except Exception as e:
        logger.error(f"Failed to get cluster slots: {e}")
        return {"status": "error", "error": "Failed to retrieve cluster slot distribution"}


@router.get("/cluster/info")
async def get_cluster_info():
    """
    Get detailed Redis cluster information

    Returns comprehensive cluster state including:
    - Cluster state (ok/fail)
    - Slots assigned/ok/fail
    - Known nodes
    - Cluster size
    - Epoch information
    - Stats
    """
    try:
        # Fetch credentials from Vault
        creds = await vault_client.get_secret("redis-1")
        password = creds.get("password")

        # Connect to first node
        client = redis.Redis(
            host=settings.REDIS_HOST,
            port=settings.REDIS_PORT,
            password=password,
            decode_responses=True,
            socket_connect_timeout=5
        )

        # Get cluster info
        cluster_info_raw = await client.execute_command("CLUSTER", "INFO")
        await client.close()

        # Parse cluster info
        cluster_info = {}
        for line in cluster_info_raw.split("\n"):
            if ":" in line:
                key, value = line.strip().split(":", 1)
                # Try to convert to int if possible
                try:
                    cluster_info[key] = int(value)
                except ValueError:
                    cluster_info[key] = value

        return {
            "status": "success",
            "cluster_info": cluster_info
        }

    except Exception as e:
        logger.error(f"Failed to get cluster info: {e}")
        return {"status": "error", "error": "Failed to retrieve cluster information"}


@router.get("/nodes/{node_name}/info")
async def get_node_info(node_name: str):
    """
    Get detailed information about a specific Redis node

    Args:
        node_name: Name of the node (redis-1, redis-2, redis-3)

    Returns detailed server information for the specified node
    """
    try:
        # Validate node name
        valid_nodes = ["redis-1", "redis-2", "redis-3"]
        if node_name not in valid_nodes:
            return {
                "status": "error",
                "error": f"Invalid node name. Must be one of: {', '.join(valid_nodes)}"
            }

        # Fetch credentials from Vault
        creds = await vault_client.get_secret("redis-1")
        password = creds.get("password")

        # Connect to the specific node
        client = redis.Redis(
            host=node_name,
            port=6379,
            password=password,
            decode_responses=True,
            socket_connect_timeout=5
        )

        # Get comprehensive server info
        info = await client.info("all")
        await client.close()

        return {
            "status": "success",
            "node": node_name,
            "info": info
        }

    except Exception as e:
        # Log error without user-controlled input to prevent log injection
        logger.error(f"Failed to get Redis node info: {e}")
        return {"status": "error", "error": "Failed to retrieve node information"}
